modules/ssl/ssl_engine_init.c

```diff
@@ -189,6 +197,122 @@ static void ssl_add_version_components(apr_pool_t *ptemp, apr_pool_t *pconf,
                  modver, AP_SERVER_BASEVERSION, incver);
 }
 
+#ifdef HAVE_OPENSSL_ECH
+/* 
+ * load any key files we find in the ECHKeyDir directory 
+ * where there are matching <name>.pub and <name>.priv files
+ * that match 
+ */
+static int load_echkeys(SSL_CTX *ctx, const char *echdir, server_rec *s, apr_pool_t *ptemp)
+{
+    /*
+     * Try load any good looking public/private ECH values found in files in that directory
+     *
+     * This code is derived from what I added to openssl s_server, which you can find
+     * in apps/s_server.c in my openssl fork, https://github.com/sftcd/openssl
+     */
+    if (echdir==NULL) {
+        ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10520)
+                "load_echkeys: no directory name - exiting");
+        return -1;
+    }
+    size_t elen=strlen(echdir);
+    if ((elen+7) >= PATH_MAX) {
+        ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10521)
+                "load_echkeys: directory name too long: %s - exiting",echdir);
+        return -1;
+    }
+    int keystried=0;
+    int keysworked=0;
+
+    apr_dir_t *dir;
+    apr_finfo_t direntry;
+    apr_int32_t finfo_flags = APR_FINFO_TYPE|APR_FINFO_NAME;
+
+    if (!echdir || (apr_dir_open(&dir, echdir, ptemp) != APR_SUCCESS)) {
+        ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10522)
+                "load_echkeys: can't open directory %s - exiting",echdir);
+        return -1;
+    }
+
+    OSSL_ECHSTORE * const es = OSSL_ECHSTORE_new(NULL, NULL);
+    if (es == NULL) {
+        ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10523)
+                "load_echkeys: can't alloc store");
+        return -1;
+    }
+
+    while ((apr_dir_read(&direntry, finfo_flags, dir)) == APR_SUCCESS) {
+        const char *fname;
+        if (direntry.filetype == APR_DIR) {
+            continue; /* don't try to load directories */
+        }
+        fname = apr_pstrcat(ptemp, echdir, "/", direntry.name, NULL);
+        /*
+         * If file name matches "*.ech" then try load that 
+         */
+        if (!fname) {
+            continue;
+        }
+        size_t pnlen=strlen(fname);
+        if (pnlen<5 || pnlen>PATH_MAX-1) {
+            continue;
+        }
+        if (!(fname[pnlen-4]=='.'
+            && fname[pnlen-3]=='e'
+            && fname[pnlen-2]=='c'
+            && fname[pnlen-1]=='h')) {
+            continue;
+        }
+        /* should likely use apr_stat instead */
+        struct stat thestat;
+        apr_finfo_t theinfo;
+        if ( (apr_stat (&theinfo, fname, APR_FINFO_MIN, ptemp)==APR_SUCCESS) ) {
+            keystried++;
+
+            BIO *in = BIO_new_file(fname, "r");
+            const int is_retry_config = OSSL_ECH_FOR_RETRY;
+            if (in != NULL
+                && 1 == OSSL_ECHSTORE_read_pem(es, in, is_retry_config)) {
+                ap_log_error(APLOG_MARK, APLOG_TRACE4, 0, s, APLOGNO(10524)
```

Comment:
For `APLOG_TRACE?` logging we don't use `APLOGNO`s.

```suggestion
                ap_log_error(APLOG_MARK, APLOG_TRACE4, 0, s,
```

Reply:
yep fixed that